# Customer Dataset Cleansing ( 4.9 Part One)

## Importing Libs

In [2]:
import pandas as pd
import numpy as np
import os

## Importing Dataset

In [3]:
# importing customers
path = r'C:\Users\shyou\3D Objects\My Original Data Sources\Achievement4\March-2023 Instacart Basket Analysis'

In [4]:
df_customers = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [5]:
df_customers

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [6]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [7]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [8]:
# importing orders_products_merged
orders_products_merge = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# Data Wrangling

### Dropping column

##### In this stage, I prefer to not drop any columns because maybe we need them in the future

### Finding missing values

In [9]:
df_customers.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [43]:
 df_nan = df_customers.loc[df_customers['First Name'].isnull() == True]

In [44]:
df_nan.head()

user_id First Name   Surnam  Gender      STATE  Age date_joined  \
53     76659        NaN  Gilbert    Male   Colorado   26    1/1/2017   
73     13738        NaN    Frost  Female  Louisiana   39    1/1/2017   
82     89996        NaN   Dawson  Female     Oregon   52    1/1/2017   
99     96166        NaN  Oconnor    Male   Oklahoma   51    1/1/2017   
105    29778        NaN   Dawson  Female       Utah   63    1/1/2017   

     n_dependants fam_status  income  
53              2    married   41709  
73              0     single   82518  
82              3    married  117099  
99              1    married  155673  
105             3    married  151819

##### We found 11259 missing values in 'First Name' column. Because the ID of the customers is used in other datasets and will be used in the analysis process, so I will not delete the rows whose first_name is Null and keep them.

### Renaming column

In [10]:
df_customers.rename(columns = {'First Name': 'first_name'}, inplace = True)

In [11]:
df_customers.rename(columns = {'Surnam': 'last_name'}, inplace = True)

In [12]:
df_customers.rename(columns = {'Gender': 'gender'}, inplace = True)

In [13]:
df_customers.rename(columns = {'STATE': 'state'}, inplace = True)

In [14]:
df_customers.rename(columns = {'Age': 'age'}, inplace = True)

In [15]:
df_customers.rename(columns = {'n_dependants': 'num_of_dependants'}, inplace = True)

In [16]:
df_customers

user_id first_name last_name  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        num_of_dependants fam_status  income  
0                       3    married  165665  
1                       0     single   59285  
2                       2    married   99568  
3                       0     single   42049  
4                       1    married   40374  
...                   ...        ...     ...  
206204                  1    married  148828  
206205                  3    married  168639  
206206                  2    married   53374  
206207                  1    married   99799  
206208                  1    married   57095  

[206209 rows x 10 columns]

### Changing data type

In [17]:
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [18]:
df_customers['user_id'].dtype

dtype('O')

# Data Consistency Checks

### Mixed-Type Data

In [19]:
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


In [20]:
df_customers.dtypes

user_id              object
first_name           object
last_name            object
gender               object
state                object
age                   int64
date_joined          object
num_of_dependants     int64
fam_status           object
income                int64
dtype: object

In [22]:
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [23]:
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

### Duplicates

In [24]:
df_customers_dups = df_customers[df_customers.duplicated()]

In [25]:
df_customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_of_dependants, fam_status, income]
Index: []

In [26]:
df_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'customers_cecked.pkl'))

## Merge Datasets

In [27]:
df_customers['user_id'] = df_customers['user_id'].astype('int64')

In [28]:
orders_products_merge = pd.read_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [29]:
orders_products_merge_2 = orders_products_merge.drop(['prior_merge','_merge', 'aisle_id','order_frequency_flag', 'add_to_cart_order', 'product_name'] , axis= 1)

In [30]:
orders_products_merge_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 21 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   product_id              int64  
 1   department_id           int64  
 2   prices                  float64
 3   order_id                int64  
 4   user_id                 int64  
 5   order_number            int64  
 6   order_day_of_week       int64  
 7   order_hour_of_day       int64  
 8   days_since_prior_order  float64
 9   reordered               int64  
 10  price_range_loc         object 
 11  busiest_day             object 
 12  busiest_days            object 
 13  busiest_period_of_day   object 
 14  max_order               int64  
 15  loyalty_flag            object 
 16  dep_mean_order          float64
 17  user_mean_payment       float64
 18  user_spender_flag       object 
 19  user_median_order       float64
 20  order_frequency_flag    object 
dtypes: float64(5), int64(9), obje

In [31]:
orders_products_customers_merge = df_customers.merge(orders_products_merge_2, on = 'user_id', indicator = True)

In [32]:
orders_products_customers_merge.head()

user_id first_name last_name  gender     state  age date_joined  \
0    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
1    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
2    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
3    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   
4    26711    Deborah  Esquivel  Female  Missouri   48    1/1/2017   

   num_of_dependants fam_status  income  ...    busiest_days  \
0                  3    married  165665  ...     Busiest day   
1                  3    married  165665  ...  Regularly busy   
2                  3    married  165665  ...     Busiest day   
3                  3    married  165665  ...  Regularly busy   
4                  3    married  165665  ...     Slowest day   

   busiest_period_of_day  max_order  loyalty_flag  dep_mean_order  \
0            Most orders          8  New customer       17.225802   
1            Most orders          8  New customer       17.225802   
2            Most orders          8  New customer       17.225802   
3            Most orders          8  New customer       17.811403   
4            Most orders          8  New customer       17.811403   

   user_mean_payment  user_spender_flag  user_median_order  \
0           7.988889        Low spender               19.0   
1           7.988889        Low spender               19.0   
2           7.988889        Low spender               19.0   
3           7.988889        Low spender               19.0   
4           7.988889        Low spender               19.0   

   order_frequency_flag  _merge  
0       Regular customer   both  
1       Regular customer   both  
2       Regular customer   both  
3       Regular customer   both  
4       Regular customer   both  

[5 rows x 31 columns]

In [33]:
orders_products_customers_merge['_merge'].value_counts(dropna = False)

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [34]:
orders_products_customers_merge.shape

(32404859, 31)

In [35]:
orders_products_customers_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_customers_merge.pkl'))